<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>INSP3CT by shayben</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>INSP3CT</h1>
        <p>Interpolation and Statistical Proximity of 3C Tables</p>

        <p class="view"><a href="https://github.com/shayben/INSP3CT">View the Project on GitHub <small>shayben/INSP3CT</small></a></p>


        <ul>
          <li><a href="https://github.com/shayben/INSP3CT/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/shayben/INSP3CT/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/shayben/INSP3CT">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h1>INSP3CT</h1>

<hr><h2>Interpolation and Statistical Proximity of 3C Tables</h2>

<h3>Version 1.0</h3>

<hr><p>Provided as supplementary alongside - </p>

<blockquote>
<p>"Spatial localization of co-regulated genes exceeds genomic gene clustering in the S. cerevisiae genome"
By Shay Ben-Elazar, Zohar Yakhini, and Itai Yanai. Accepted for publication to NAR (Dec 2012).</p>
</blockquote>

<hr><h2>Description</h2>

<p>This program can be used to take a sparse contact dataset and interpolate it to a sampled matrix at an arbitrary resolution. On top of the resulting matrix, any set of annotation of loci is then inspected for co-localization both in 3d and 1d.
The result can then optionally be displayed in a figure similar to Figure 3 in the above paper.</p>

<h2>Requirements</h2>

<p>It is recommended to run this on a machine with at least 8 GB of RAM, depending on the size of the dataset.
Multiple cores can and should be utilized to speed up analysis.
Our setup (for this particular dataset) was a machine with 192 GB RAM and 12 dedicated cores.
Basic Matlab knowhow is a prerequisite.</p>

<h2>Input and parameters:</h2>

<p>Edit the main pipeline file, INSP3CT.m, and make sure to set the following parameters.</p>

<h3>Required input files:</h3>

<ul>
<li>PARAMETERS.interfrequencies - filename of the inter-chromosomal frequencies tab-delimited in the format with the following header:<br>
chr1    locus1  chr2    locus2  frequency</li>
<li>PARAMETERS.intrafrequencies - filename of the intra-chromosomal frequencies tab-delimited in the same format, where chr1=chr2.</li>
<li>PARAMETERS.ChromosomeSizes - An array holding the number of bases in each chromosome in the genome.</li>
</ul><p>An example of a sufficient file-set can be found on github (<a href="https://github.com/shayben/INSP3CT/blob/master/YeastDatasetExample.zip">YeastDatasetExample.zip</a>). The parameters corresponding  to this input are commented out in the code.</p>

<h3>Additional parameters:</h3>

<ul>
<li>PARAMETERS.cores - number of cores in the machine used for processing. More cores is faster but requires more RAM as well.</li>
<li>PARAMETERS.binningresolution - the number of nucleotides which are considered as a locus. Smaller value -&gt; larger dataset and slower to compute but provides more accuracy in the results.</li>
<li>PARAMETERS.analysis_shuffle - set to true for a random control.</li>
<li>PARAMETERS.analysis_cyclic - set to true for a cyclic permutation control.</li>
</ul><h3>You also need to manually create the following variables before running any co-localization analysis:</h3>

<ul>
<li>Loci - A vector of k bin indices containing points of interest.</li>
<li>AnnotationData - A matrix of size kXn where each row is a point of interest and each column an annotation. The (i,j)-th entry is the number of elements annotated as j which reside in bin Loci[i].</li>
</ul><p>There is a commented-out example which randomly generates loci and annotations used for testing.</p>

<h2>Usage</h2>

<p>Main file is INSP3CT.m.</p>

<ul>
<li>The first code block, labeled "INSP3CT pipeline", loads the parameters and asserts their validity.</li>
<li>The second code block, labeled "Load and parse datasets", Prepares the data structures used to generate the interpolated contact frequency matrix.</li>
<li>The third code block, labeled "Generate interpolated and binned frequencies", runs the interpolation step.
### The next two blocks can be used if Loci and AnnotationData are set:</li>
<li>The forth code block, labeled "Analysis on annotation" given an annotation on the binned genome, analyzes 3D vs 1D co-localizations using mHG (1,2).</li>
<li>The fifth code block, labeled "Show a result figure for each annotation", plots each annotation showing the resulting landscape of enrichment. Pressing any keyboard key will continue to show the next annotation.</li>
</ul><blockquote>
<ol>
<li>Eden, E., Lipson, D., Yogev, S. and Yakhini, Z. (2007) Discovering motifs in ranked lists of DNA sequences. PLoS Comput Biol, 3, e39 </li>
<li>Eden, E., Navon, R., Steinfeld, I., Lipson, D. and Yakhini, Z. (2009) GOrilla: a tool for discovery and visualization of enriched GO terms in ranked gene lists. BMC Bioinformatics, 10, 48.).</li>
</ol>
</blockquote>

<h2>Output</h2>

<ul>
<li>obj.Interpolated - cell array containing interpolated chromosomal contact frequencies binned to the requested resolution.
### If Loci and AnnotationData are set </li>
<li>The resulting enrichment comparisons are a set of figures for the respective annotations.</li>
<li>corrected3d - Corrected mHG (partition limited) score for 3D enrichment around loci for each annotation.</li>
<li>corrected1d - Corrected mHG (partition limited) score for 1D enrichment around loci for each annotation.</li>
<li>allbestspheresizes - mHG threshold on 3D ordered loci.</li>
<li>allbestintsizes - mHG threshold on 1D ordered loci.</li>
</ul><h2>Disclaimer</h2>

<p>These scripts are provided as is and under the GPL license.
Much of the code is <em>NOT</em> optimized since it has undergone many iterations and was not specially tailored for the purposes in this code. Please make sure to read the instructions and paper thoroughly before contacting the authors.
Importantly, please note that the analysis in the paper was performed on a haploid genome and there are repercussions for switching to a diploid genome in the interpretation of the input data which were not addressed in this code or paper.</p>

<h2>Copyright</h2>

<blockquote>
<p>This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.</p>

<p>This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.</p>

<p>You should have received a copy of the GNU General Public License
along with this program.  If not, see <a href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>.</p>
</blockquote>

<h2>Contact</h2>

<p>For questions regarding this program e-mail: shayben@cs.technion.ac.il.
Further inqueries can be directed to Itai Yanai, <a href="http://yanailab.technion.ac.il/">website</a> email: <a href="mailto:yanai@technion.ac.il">yanai@technion.ac.il</a>.</p>

<hr><p>Copyright 2012 Shay Ben Elazar Â©</p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/shayben">shayben</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-36841815-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>